core/schema/function.graphqls

```diff
@@ -0,0 +1,126 @@
+scalar FunctionID
+
+extend type Query {
+  "Load a function by ID"
+  function(id: FunctionID!): Function!
+}
+
+type Function {
+  id: FunctionID!
+
+  "The name of the function"
+  name: String!
+
+  "A doc string for the function, if any"
+  description: String
+
+  "Arguments accepted by this function, if any"
+  args: [FunctionArg!]
+
+  "The type returned by this function"
+  returnType: TypeDef!
+
+  """
+  Execute this function using dynamic input+output types.
+
+  Typically, it's preferable to invoke a function using a type
+  safe graphql query rather than using this call field. However,
+  call is useful for some advanced use cases where dynamically
+  loading arbitrary modules and invoking functions in them is
+  required.
+  """
+  call(input: [FunctionCallInput!]): JSON!
+}
+
+type FunctionArg {
+  "The name of the argument"
+  name: String
+
+  "A doc string for the argument, if any"
+  description: String
+
+  "The type of the argument"
+  typeDef: TypeDef!
+
+  "A default value to use for this argument if not explicitly set by the caller, if any"
+  defaultValue: JSON
+}
+
+input FunctionCallInput {
+  "The name of the argument to the function"
+  name: String!
+
+  """
+  The value of the argument to the function, representing as a JSON-serialized string.
+  """
+  value: JSON!
+}
+
+"""
+A definition of a parameter or return type in a Module.
+"""
+type TypeDef {
```

Comment:
This could look a bit nicer with SDK support for union types, but I don't think it's realistic to rope in updating every SDK to support that as part of this effort. It also would likely be a fairly marginal improvement rather than dramatic.